design-proposals/klip-24-key-column-semantics-in-queries.md

```diff
@@ -0,0 +1,506 @@
+
+The differences with joins is more subtle.
+
+With a join query, a transient query selects all columns from all sources. A persistent query adds
+all columns from all sources, and also adds an additional column that stores the result of the join
+criteria.
+
+```sql
+SELECT * FROM INPUT I1 JOIN INPUT_2 I2 ON I1.ID = I2.ID;
+-- resulting columns: I1_ID INT, I1_V0 INT, I1_V1 INT, I2_ID INT, I2_V0 INT, I2_V1 INT
+
+-- vs --
+
+CREATE TABLE OUTPUT AS SELECT * FROM INPUT I1 JOIN INPUT_2 I2 ON I1.ID = I2.ID;
+-- resulting schema (any key name enabled): ID INT KEY, I1_ID INT, I1_V0 INT, I1_V1 INT, I2_ID INT, I2_V0 INT, I2_V1 INT
+-- resulting schema (any key name disabled): ROWKEY INT KEY, I1_ID INT, I1_V0 INT, I1_V1 INT, I2_ID INT, I2_V0 INT, I2_V1 INT
+```
+
+Note the addition of an additional `ID` or `ROWKEY` column in the case of the persistent query.
+
+#### Inner and left outer joins on column references
+
+For inner and left outer joins where the left join criteria is a column reference the the
+additional `ID` column is a duplicate of the `I1.ID` column.
+
+We propose that such joins should not duplicate the left join column `I1.ID` into both the `ID` key
+and `I1_ID` value column. Instead, the key column should be named `I1_ID` and no copy should be stored
+in the value.
+
+#### Full outer joins and non-column joins
+
+Where the join is a full outer join, or where the left join criteria is not a column reference, the
+problem is more nuanced.
+
+```sql
+-- inner join on expression:
+CREATE TABLE OUTPUT AS SELECT * FROM INPUT I1 JOIN INPUT_2 I2 ON ABS(I1.ID) = I2.ID;
+
+-- full outer join:
+CREATE TABLE OUTPUT AS SELECT * FROM INPUT I1 OUTER JOIN INPUT_2 I2 ON I1.ID = I2.ID;
+```
+
+For both of the above joins the data stored in the Kafka record's key by ksqlDB / Streams does
+not correspond to any column within either source. This is problematic.
+
+This KLIP proposes that the projection should include _all the columns expected in the result_.
+Logically, this must include this new key column. Yet, this new key column is an artifact of the
+current join implementation, so what should it be called and how should the user include it in the
+projection?
+
+The synthesised column is currently named `ROWKEY`. However, the 'any key name' feature removes the
+`ROWKEY` system column in favour of user supplied names or system generated ones in the form
+`KSQL_COL_x`.
+
+If we are to require users to explicitly include this synthesised column in any projection with
+explicit columns, i.e. non select-star projections, then the user must be able to determine the name
+and be able to provide their own. This precludes the user of a system generated name such as
+`KSQL_COL_0`.
+
+Though not ideal, we propose that in the short term the synthesised column will be named `ROWKEY`,
+as this is a name users are already familiar with. For example, the above examples that used `*` in
+their projections could be expanded to the following explicit column lists:
```

Comment:
I'm still not convinced about the use of `ROWKEY` for the synthesised column in some joins. I'm currently coding this up and will play around to see what works best. 

One concern with `ROWKEY` is that a lot of existing users will have sources with a `ROWKEY` key column, meaning there'll be a column clash, which is nasty.

Alternatives are:

* a new reserved `JOINKEY` column name.

* system generated `KSQL_COL_x` style name, which would be inline with the proposed `GROUP BY` implementation when grouping on a non-column expression, e.g. 

```sql
CREATE TABLE OUTPUT AS
   SELECT KSQL_COL_0, COUNT() FROM INPUT GROUP BY MyUdf(V0);
````

The difference here is that the grouping term `MyUdf(V0)` obviously creates a new column and its named the same as if the user had done:

```sql
SELECT MyUdf(V0) FROM INPUT;
-- resulting columns: KSQL_COL_0.
```

Where as the join has a new column that's purely an artifact of our join impl... :-/

Reply:
I've concluded my investigation into this and the easiest solution to implement, and the one that avoids nasty edge cases, is the `JOINKEY()` udf.

I've updated the KLIP to refelect this.

Reply:
I am not particular happy to require users to call an UDF to make the query work... That seems even worse than requiring them to include the PK column in the `SELECT` clause.

I understand the motivation (ie, the complexity to use `ROWKEY` or similar rejected alternatives); however, I am a little worried about the user experience. To what extend do we need to carry the burden and to what extend can/shall we put it into the users?

Reply:
Though I agree this is not ideal, I can't think of a workable alternative in the short term.  

Deadlines being what they are, this code is already implemented and merged.  Though not documented yet.  Can we agree this is not ideal, but can always be improved in a later KLIP

Reply:
> this code is already implemented and merged

_and merged_ -- the KLIP is not finalized yet? Interesting "process".

Reply:
Lol, indeed.

Been thinking on this. Totally agree JOINKEY is horrible. My plan is to get everything else merged and turn on, and `WITH(KEY)` removed. This will massively simplify the code.  With that done, I can spend some time to see how things fall out if we remove the requirement for JOINKEY, but allow JOINKEY is the user wants.

Reply:
Well. Leave it up to you.

Reply:
Issue that tracks getting rid of JoinKey: https://github.com/confluentinc/ksql/issues/5304